darts/models/transformer_model.py

```diff
@@ -0,0 +1,281 @@
+        # see section 3.2.1 in 'Attention is All you Need' by Vaswani et al. (2017)
+        src = self.encoder(src) * math.sqrt(self.input_size)
+        src = self.positional_encoding(src)
+
+        tgt = self.encoder(tgt) * math.sqrt(self.input_size)
+        tgt = self.positional_encoding(tgt)
+
+        x = self.transformer(src=src,
+                             tgt=tgt)
+        out = self.decoder(x)
+
+        # Here we change the data format
+        # from (1, batch_size, output_length * output_size)
+        # to (batch_size, output_length, output_size)
+        predictions = out[0, :, :]
+        predictions = predictions.view(-1, self.output_length, self.output_size)
+
+        return predictions
+
+
+class TransformerModel(TorchForecastingModel):
+    @random_method
+    def __init__(self,
+                 model: nn.Module = None,
+                 input_size: int = 1,
+                 input_length: int = 1,
+                 output_length: int = 1,
+                 output_size: int = 1,
+                 d_model: int = 512,
+                 nhead: int = 8,
+                 num_encoder_layers: int = 6,
+                 num_decoder_layers: int = 6,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 activation: str = "relu",
+                 custom_encoder: Optional[Any] = None,
+                 custom_decoder: Optional[Any] = None,
+                 random_state: Optional[Union[int, RandomState]] = None,
+                 **kwargs):
+
+        """
+        Transformer is a state-of-the-art deep learning model introduced in 2017. It is an encoder-decoder
+        architecture whose core feature is the 'multi-head attention' mechanism, which is able to
+        draw intra-dependencies within the input vector and within the output vector ('self-attention')
+        as well as inter-dependencies between input and output vectors ('encoder-decoder attention').
+        The multi-head attention mechanism is highly parallelizable, which makes the transformer architecture
+        very suitable to be trained with GPUs.
+
+
+        The transformer architecture implemented here is based on the paper “Attention Is All You Need”:
+        Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan N Gomez, Lukasz Kaiser, and
+        Illia Polosukhin. 2017. Attention is all you need. In Advances in Neural Information Processing Systems,
+        pages 6000-6010.
+        (paper can be found at https://arxiv.org/abs/1706.03762)
+
+        Parameters
+        ----------
+        model
+            a custom PyTorch module with the same specifications as
+            `darts.models.transformer_model._TransformerModule` (default=None).
+        input_size
+            The dimensionality of the TimeSeries instances that will be fed to the fit function (default=1).
+        input_length
+            Number of time steps to be output by the forecasting module (default=1).
```

Comment:
This is inaccurate (we also have this typo, which we should eventually correct, in the RNN docstring). It's the number of steps that will be input to the transformer module.